src/hotspot/share/c1/c1_GraphBuilder.cpp

```diff
@@ -2027,8 +2027,10 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
       int index = state()->stack_size() - (target->arg_size_no_receiver() + 1);
       receiver = state()->stack_at(index);
       ciType* type = receiver->exact_type();
-      if (type != nullptr && type->is_loaded() &&
-          type->is_instance_klass() && !type->as_instance_klass()->is_interface()) {
+      if (type != nullptr && type->is_loaded()) {
```

Comment:
Is it the case that we can't see an interface here? Or that we think it's ok if we see an interface here?

Reply:
We can't see an interface here because it will get rejected by `ciInstanceKlass::exact_klass`, so we could even assert for that here if we wanted.

Reply:
Then, I think we should add an assert that `!type->as_instance_klass()->is_interface()` and also that it's not and array of interfaces (using `base_element_klass()`)

Reply:
An array of interfaces can be exact:
```
    new Interface[20].getClasss();
```
and it seems like it would be safe to allow this, so I think we only need one assert for `!type->as_instance_klass()->is_interface()` if we don't trust the result of exact_type().

Reply:
Right. Then I think it would be safer to add an assert for `!type->as_instance_klass()->is_interface()`.

Reply:
@dean-long @rwestrel I've added the assert.

Reply:
The assert doesn't hold, e.g.

```
=== Output from failing command(s) repeated here ===
* For target buildtools_create_symbols_javac__the.COMPILE_CREATE_SYMBOLS_batch:
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/runner/work/jdk/jdk/src/hotspot/share/c1/c1_GraphBuilder.cpp:2031), pid=75212, tid=75244
#  Error: assert(!!type->as_instance_klass()->is_interface()) failed
#
# JRE version: OpenJDK Runtime Environment (23.0) (fastdebug build 23-internal-galderz-306db7459b1316251e36d0eccc3035d11db44889)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 23-internal-galderz-306db7459b1316251e36d0eccc3035d11db44889, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x777cd0]  GraphBuilder::invoke(Bytecodes::Code)+0x1200
```

Thoughts @rwestrel @dean-long?

Reply:
Hmmm, the double `!!`... let me fix that and see.

Reply:
Hmmm, something else is failing now. That's odd, maybe master has updated and is causing this PR to fail now?

```
#  Internal Error (/Users/runner/work/jdk/jdk/src/hotspot/share/ci/ciMetadata.hpp:88), pid=79328, tid=27395
#  assert(is_instance_klass()) failed: bad cast
```

I will look into it.

Reply:
Ah no, that assert comes from `type->as_instance_klass()` call:

```
  ciInstanceKlass*         as_instance_klass() {
    assert(is_instance_klass(), "bad cast");
    return (ciInstanceKlass*)this;
  }
```

@rwestrel @dean-long what shall we do here? Do we remove the assert altogether? Does the code need to change for the assert to pass? Any other ideas?